internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -0,0 +1,499 @@
+}
+
+// Finds the agents related to this BMH
+//
+// The BMH <-> agent relation should be one-to-one. This
+// function returns the agent that matches the following
+// criteria.
+//
+// An agent is only related to the BMH if it has an iface
+// whose MacAddress matches the BootMACAddress set in the BMH.
+//
+// In the scenario where more than one Agent match the BootMACAdress
+// from the BMH, the following logic applies:
+//
+// 1. Return the agent that has been approved, if any.
+// 2. If the above fails, return the first agent found.
+//
+// Otherwise, this function will return nil.
+//
+// Eventually, this will be changed to inspect the agents status
+// instead of `Approved`. We want to reconcile only agents that
+// are in specific states (like discovering). Waiting on MGMT-4266
+func (r *BMACReconciler) findAgent(ctx context.Context, bmh *bmh_v1alpha1.BareMetalHost) *adiiov1alpha1.Agent {
+	agentList := adiiov1alpha1.AgentList{}
+	err := r.Client.List(ctx, &agentList, client.InNamespace(bmh.Namespace))
+	if err != nil {
+		return nil
+	}
+
+	agents := []*adiiov1alpha1.Agent{}
+
+	for i, agent := range agentList.Items {
+		for _, agentInterface := range agent.Status.Inventory.Interfaces {
+			if bmh.Spec.BootMACAddress == agentInterface.MacAddress {
+
+				// Short circuit if we find an agent
+				// that is Approved
+				if agent.Spec.Approved {
+					return &agent
+				}
+
+				agents = append(agents, &agentList.Items[i])
+			}
+		}
+	}
+
+	if len(agents) > 0 {
+		return agents[0]
+	}
+
+	return nil
+}
+
+// Find `BareMetalHost` resources that match an agent
+//
+// Only `BareMetalHost` resources that match one of the Agent's
+// MAC addresses will be returned.
+func (r *BMACReconciler) findBMH(ctx context.Context, agent *adiiov1alpha1.Agent) (*bmh_v1alpha1.BareMetalHost, error) {
+	bmhList := bmh_v1alpha1.BareMetalHostList{}
+	err := r.Client.List(ctx, &bmhList, client.InNamespace(agent.Namespace))
+	if err != nil {
+		return nil, err
+	}
+
+	for _, bmh := range bmhList.Items {
+		for _, agentInterface := range agent.Status.Inventory.Interfaces {
+			if bmh.Spec.BootMACAddress == agentInterface.MacAddress {
+				return &bmh, nil
+			}
+		}
+	}
+	return nil, nil
+}
+
+func (r *BMACReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	mapSecretToClusterDeployment := handler.ToRequestsFunc(
+		func(a handler.MapObject) []reconcile.Request {
+			ctx := context.Background()
+			agent := &adiiov1alpha1.Agent{}
+
+			if err := r.Get(ctx, types.NamespacedName{Name: a.Meta.GetName(), Namespace: a.Meta.GetNamespace()}, agent); err != nil {
+				return []reconcile.Request{}
+			}
+
+			// No need to list all the `BareMetalHost` resources if the agent
+			// already has the reference to one of them.
+			if val, ok := agent.ObjectMeta.Labels[AGENT_BMH_LABEL]; ok {
+				return []reconcile.Request{{NamespacedName: types.NamespacedName{
+					Namespace: agent.Namespace,
+					Name:      val,
+				}}}
+			}
+
+			bmh, err := r.findBMH(ctx, agent)
```

Comment:
As i understand, say me of i am wrong.
We get event on agent cr, we find relevant bmh and from this point we run reconcile on this bmh? One of the functions we run on bmh reconciliation is findAgent. So why we just don't relevant agent from here?

Reply:
The mapping is only giving you an even that some object is need to be reconciled, it does not pass additional information.

Reply:
got it, pity though